tower-http/src/validate_request.rs

```diff
@@ -50,6 +52,70 @@
 //! # }
 //! ```
 //!
+//! Validation of a custom header can be made by using [`ValidateRequestHeaderLayer::assert()`]:
+//!
+//! ```
+//! use tower_http::validate_request::ValidateRequestHeaderLayer;
+//! use hyper::{Request, Response, Body, Error};
+//! use http::StatusCode;
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let mut service = ServiceBuilder::new()
+//!     // Require a `X-Custom-Header` header to have the value `random-value-1234567890` or reject with a `403 Forbidden` response
+//!     .layer(ValidateRequestHeaderLayer::assert("x-custom-header", "random-value-1234567890", StatusCode::FORBIDDEN))
```

Comment:
I also though about using a Builder pattern here

so instead of the `assert` function, the API would look like:

```rs
.layer(
    ValidateRequestHeaderLayer::has("x-custom-header").with_value("random-value-1234567890").or_reject_with(StatusCode::FORBIDDEN)
)
```

where `or_reject_with` would be the "build" method...
and we could say that `with_value` is optional, so this new layer could also be used by those who just want to make sure that a request has a specific header but we don't care about its value